src/rust/rust-proto/src/graplinc/grapl/common/v1beta1/types.rs

```diff
@@ -0,0 +1,245 @@
+use crate::{
+    protobufs::graplinc::grapl::common::v1beta1::{
+        EdgeName as EdgeNameProto,
+        NodeType as NodeTypeProto,
+        PropertyName as PropertyNameProto,
+        Uid as UidProto,
+    },
+    serde_impl,
+    type_url,
+    SerDeError,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PropertyName {
+    pub value: String,
+}
+
+impl std::fmt::Display for PropertyName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.value)
+    }
+}
+
+impl PropertyName {
+    pub fn new_unchecked(value: String) -> Self {
+        // todo: debug assertion
+        Self { value }
+    }
+}
+
+impl TryFrom<&'static str> for PropertyName {
+    type Error = SerDeError;
+    fn try_from(raw: &'static str) -> Result<Self, Self::Error> {
+        if raw.is_empty() {
+            return Err(SerDeError::InvalidField {
+                field_name: "PropertyName",
+                assertion: "can not be empty".to_owned(),
+            });
+        }
+        if raw.len() > 32 {
+            return Err(SerDeError::InvalidField {
+                field_name: "PropertyName",
+                assertion: "can not be more than 32 characters".to_owned(),
+            });
+        }
+
+        Ok(Self {
+            value: raw.to_owned(),
+        })
+    }
+}
+
+impl TryFrom<String> for PropertyName {
+    type Error = &'static str;
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        // todo: Validate this thing
+        Ok(PropertyName { value })
```

Comment:
Does the above `TryFrom` _have_ to be on a `&'static str`, or can it be just on a `&str`? If the latter, you could use the same validation logic here.

Reply:
Or maybe you can recast the validation logic to be in terms of an `AsRef<str>`... then you could externalize it and share it across all the conversion functions for this and anything else that has to be a non-empty, less-than-32-character string.

Reply:
Do you mean does the error have to be &'static? If so, it either has to be static or a String, since errors are returned all the way up through main so they must always be `'static`.

Reply:
Great call, I made it a simple
```
impl TryFrom<&str> for EdgeName {
    type Error = SerDeError;
    fn try_from(raw: &str) -> Result<Self, Self::Error> {
        EdgeName::try_from(raw.to_owned())
    }
}
```

Reply:
@colin-grapl I'm not talking about the error type here; I'm talking about the source type for the conversion (`TryFrom<&'static str> for PropertyName` vs. `TryFrom<String> for PropertyName`). I think you could just have this be `TryFrom<AsRef<str>> for PropertyName` (or whatever the syntax / type parameterization is). You could also externalize the validation logic to be reused across all these string-like things that have to be non-empty and less than 32 characters.

Reply:
Ah ok. In that case I would suggest two implementations, one for String, one for &str. It doesn't have to be static, you're correct.

Reply:
The reason is because PropertyName requires an owned String, so taking AsRef<str> would require cloning even when the caller can provide a String, which is the common case.